astroid/brain/brain_builtin_inference.py

```diff
@@ -481,6 +481,99 @@ def infer_slice(node, context=None):
     return slice_node
 
 
+def infer_open(node, context=None):
+    """Understand `open` (and `file` in Python 2) calls."""
+    open_return_classdef_node = _get_open_return_classdef_node(node, context)
+    return open_return_classdef_node.instantiate_class()
+
+
+def _get_open_return_classdef_node(node, context):
+    if six.PY2:
+        node = _get_py2_open_return_class(node, context)
+    else:
+        node = _get_py3_open_return_class(node, context)
+    return next(node.infer())
+
+
+def _get_py2_open_return_class(node, context):  # pylint: disable=unused-argument
+    code = dedent('''
+        file #@
+    ''')
+    return extract_node(code)
+
+
+def _get_py3_open_return_class(node, context):
+    code = '''
+        import io
+        io.{class_name}  #@
+    '''
+    class_name = _get_py3_open_class_name(node, context)
+    return extract_node(code.format(class_name=class_name))
+
+
+def _get_py3_open_class_name(node, context):
+    # https://docs.python.org/3/library/functions.html#open
+    # TODO: use call_site
+    mode = _get_mode(node, context)
+    buffering = _get_buffering(node, context)
+
+    if not buffering:
+        return 'FileIO'
+
+    if 'b' not in mode:
+        return 'TextIOWrapper'
+    else:
+        if '+' in mode:
+            return 'BufferedRandom'
+        elif 'w' in mode or 'a' in mode:
+            return 'BufferedWriter'
+        else:
+            return 'BufferedReader'
+
+
+def _get_mode(node, context):
+    return _get_call_arg(node, context, arg=1, kwarg='mode', default='r')
+
+
+def _get_buffering(node, context):
+    return _get_call_arg(node, context, arg=2, kwarg='buffering', default=-1)
+
+
+def _get_call_arg(node, context, arg, kwarg, default):
+    """TODO: move to generic solution"""
+    if len(node.args) > arg:
+        inferred = helpers.safe_infer(node.args[arg], context)
+        if inferred:
+            return inferred.value
+    kws = [kw for kw in node.keywords or [] if kw.arg == kwarg]
+    if len(kws) == 1:
+        inferred = helpers.safe_infer(kws[0].value, context)
+        if inferred:
+            return inferred.value
+    return default
+
+
+def enter_returns_self(node):
```

Comment:
I think this is (or may be) a generic operation. We have a "dummy" `ClassDef` built based on C-implemented built-ins where `FunctionDef` in bodies are empty.

General task would be to patch instances (based on predicate)  - based on some fake node. Key statement here is to _patch_ selectively, not to _replace_ with stub.